File: Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(nums,[],res)
        return res
        
    def dfs(self,nums,path,res):
        if len(nums) == 0:
            res.append(list(path))
        for i in xrange(len(nums)):
            self.dfs(nums[:i]+nums[i+1:],path+[nums[i]],res)
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
    #     res = []
    #     self.dfs(nums,[],res)
    #     return res
        
    # def dfs(self,nums,path,res):
    #     if not nums:
    #         res.append(path)
    #     for i in xrange(len(nums)):
    #         self.dfs(nums[:i] + nums[i+1:],path+[nums[i]],res)
            

File: permutations.kt
class Solution {
    var n = intArrayOf()
    var result = mutableListOf<List<Int>>()

    fun permute(nums: IntArray): List<List<Int>> {
        this.n = nums
        val picked = HashSet<Int>()
        var temp = mutableListOf<Int>()
        dfs(temp, picked)
        return this.result.toList()

    }

    fun dfs(temp: MutableList<Int>, picked: HashSet<Int>) {
        if (temp.size == this.n.size) {
            this.result.add(temp.toList())
            return
        }
        for (num in n){
            if (num !in picked){
                temp.add(num)
                picked.add(num)
                dfs(temp, picked)
                temp.remove(num)
                picked.remove(num)
            }
        }
    }
}
